[jit] Fix circular deps in loading

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26788 [jit] clean up exported source format
* #26787 [jit] Make PythonPrint a class
* #26786 [jit] remove dead LEGACY_PythonPrint
* #26767 [jit] Remove attempToRecoverType
* #26572 [jit] Allow 'Any'  to appear as a type argument.
* #26499 [jit] refactor tryMatchSchema
* #26271 [JIT] add type refinements for isinstance checks
* #26270 [jit] refactor extra sugared values
* #26269 [jit] add dynamic isinstance
* #26268 Allow types as node attributes
* **#26758 [jit] Fix circular deps in loading**

This PR changes the order in which we import classes and functions so
that is is no longer necessary for them to defined in order in a file,
or for there to be proper import statements in the exported file.

Actually importing a function/class now is driven by the need to resolve
the entity during unpickling, type resolution, or value resolution.

While this should allow significant simplification to the code that
serializes classes, this work has not been done yet in order to avoid
inevitable forward compat issues in the transition period.

Notes:
* Individual functions have been replaced with a SourceImporter object
  that exposes a resolveType method. This method loads the type if
  it has not been loaded yet, potentially parsing  (but not loading)
  the file it exists in if that file hasn't been parsed yet.
* Some legacy functionality needed to be added as a method to this object
  since the old format still used some of this logic for class resolution.

Differential Revision: [D17558989](https://our.internmc.facebook.com/intern/diff/D17558989)

zdevito committed:
[jit] Fix circular deps in loading

This PR changes the order in which we import classes and functions so
that is is no longer necessary for them to defined in order in a file,
or for there to be proper import statements in the exported file.

Actually importing a function/class now is driven by the need to resolve
the entity during unpickling, type resolution, or value resolution.

While this should allow significant simplification to the code that
serializes classes, this work has not been done yet in order to avoid
inevitable forward compat issues in the transition period.

Notes:
* Individual functions have been replaced with a SourceImporter object
  that exposes a resolveType method. This method loads the type if
  it has not been loaded yet, potentially parsing  (but not loading)
  the file it exists in if that file hasn't been parsed yet.
* Some legacy functionality needed to be added as a method to this object
  since the old format still used some of this logic for class resolution.

suo commented on torch/csrc/jit/import_source.cpp:
```diff
@@ -111,20 +95,17 @@ struct ConstantTableValue : public SugaredValue {
-      size_t version,
-      const std::vector<at::Tensor>& tensor_table)
-      : cu_(std::move(cu)) {
+struct SourceImporterImpl : public Resolver,
```
Can we name this SourceResolver? Right now seems like a pimpl thing but it also gets passed around independently as a a resolver, which is confusing.

suo commented on torch/csrc/jit/import_source.cpp:
```diff
@@ -416,58 +430,79 @@ struct SourceImporter {
+  std::unordered_set<std::string> loaded_sources_;
+  // named types and functions loaded from a file but not yet defined because
+  // their type has not been requested yet.
+  std::unordered_map<std::string, TreeRef> to_be_defined_;
```
do we define hash for QualifiedName? If so, we can make this a map of QualifiedName directly instead of `std::string`. I think it would make the code clearer

zdevito replied:
Looks like we do, so I can change this.

suo commented on torch/csrc/jit/import_source.h:
```diff
@@ -10,28 +10,35 @@
+// the source files as necessary.
+using SourceLoader = std::function<std::shared_ptr<Source>(const std::string&)>;
+
+struct TORCH_API SourceImporter {
```
Can we just have `import.cpp` use a `Resolver` directly instead of having a SourceImporter? Could be nice as it would allow us to drop a concept.

zdevito replied:
That is true, there is some weird API differences though: resolver takes a string and a source location. Providing the source location would be hard. I can yak shave further and change the resolver API though.

suo commented:
lgtm! Excited to see all the old stuff deleted 😛. 

Let's try to land this ASAP, I will add it to the 1.3 milestone.

zdevito commented on torch/csrc/jit/script/resolver.h:
```diff
@@ -34,13 +34,12 @@ struct Resolver {
       const std::string& name,
       Function& m,
-      const SourceRange& loc) const {
+      const SourceRange& loc) {
```
I think it is reasonable that a Resolver may change its internal state when resolveValue/resolveType is called. For SourceImporter, they are not constant: the importer is parsing and processing files and definitions when this stuff is called, changing its internal state. I removed it here rather than const-casting in source importer because I think this API more accurately reflects the contract that a resolver can be stateful.

zdevito commented:
I'll see if I can reorder this in my stack, but I seem to remember that it does conflict with a change I made in an earlier PR.

zdevito committed:
Update on "[jit] Fix circular deps in loading"

[jit] Fix circular deps in loading

This PR changes the order in which we import classes and functions so
that is is no longer necessary for them to defined in order in a file,
or for there to be proper import statements in the exported file.

Actually importing a function/class now is driven by the need to resolve
the entity during unpickling, type resolution, or value resolution.

While this should allow significant simplification to the code that
serializes classes, this work has not been done yet in order to avoid
inevitable forward compat issues in the transition period.

Notes:
* Individual functions have been replaced with a SourceImporter object
  that exposes a resolveType method. This method loads the type if
  it has not been loaded yet, potentially parsing  (but not loading)
  the file it exists in if that file hasn't been parsed yet.
* Some legacy functionality needed to be added as a method to this object
  since the old format still used some of this logic for class resolution.

gh-metadata: pytorch pytorch 26758 gh/zdevito/114/head

zdevito committed:
Update on "[jit] Fix circular deps in loading"

[jit] Fix circular deps in loading

This PR changes the order in which we import classes and functions so
that is is no longer necessary for them to defined in order in a file,
or for there to be proper import statements in the exported file.

Actually importing a function/class now is driven by the need to resolve
the entity during unpickling, type resolution, or value resolution.

While this should allow significant simplification to the code that
serializes classes, this work has not been done yet in order to avoid
inevitable forward compat issues in the transition period.

Notes:
* Individual functions have been replaced with a SourceImporter object
  that exposes a resolveType method. This method loads the type if
  it has not been loaded yet, potentially parsing  (but not loading)
  the file it exists in if that file hasn't been parsed yet.
* Some legacy functionality needed to be added as a method to this object
  since the old format still used some of this logic for class resolution.

gh-metadata: pytorch pytorch 26758 gh/zdevito/114/head

zdevito committed:
Update on "[jit] Fix circular deps in loading"

[jit] Fix circular deps in loading

This PR changes the order in which we import classes and functions so
that is is no longer necessary for them to defined in order in a file,
or for there to be proper import statements in the exported file.

Actually importing a function/class now is driven by the need to resolve
the entity during unpickling, type resolution, or value resolution.

While this should allow significant simplification to the code that
serializes classes, this work has not been done yet in order to avoid
inevitable forward compat issues in the transition period.

Notes:
* Individual functions have been replaced with a SourceImporter object
  that exposes a resolveType method. This method loads the type if
  it has not been loaded yet, potentially parsing  (but not loading)
  the file it exists in if that file hasn't been parsed yet.
* Some legacy functionality needed to be added as a method to this object
  since the old format still used some of this logic for class resolution.

gh-metadata: pytorch pytorch 26758 gh/zdevito/114/head

zdevito committed:
Update on "[jit] Fix circular deps in loading"

[jit] Fix circular deps in loading

This PR changes the order in which we import classes and functions so
that is is no longer necessary for them to defined in order in a file,
or for there to be proper import statements in the exported file.

Actually importing a function/class now is driven by the need to resolve
the entity during unpickling, type resolution, or value resolution.

While this should allow significant simplification to the code that
serializes classes, this work has not been done yet in order to avoid
inevitable forward compat issues in the transition period.

Notes:
* Individual functions have been replaced with a SourceImporter object
  that exposes a resolveType method. This method loads the type if
  it has not been loaded yet, potentially parsing  (but not loading)
  the file it exists in if that file hasn't been parsed yet.
* Some legacy functionality needed to be added as a method to this object
  since the old format still used some of this logic for class resolution.

gh-metadata: pytorch pytorch 26758 gh/zdevito/114/head
